<a id='top'></a>

# Introduction to Python Programming
##### Notebook to get starting to programming using Python

### By [Edd Webster](https://www.twitter.com/eddwebster)
Notebook first written: 17/06/2021<br>
Notebook last updated: 17/06/2021

![title](../../../img/python.png)

---

## 1. Introduction to Python
I'm going to be straight with you, this notebook is a little dry and quite long and one. If you've never used Python before, reading this post and understanding all the content is going to take several days/weeks (around 5-10 hours estimated, but this can vary, so don't compare yourself to others).

But have hope! By the time you've managed to work through this notebook and complete all the examples, you will have learnt a real, employable skill and gained enough of a understooding of Python that will allow you to get started with data analysis, specifically the [pandas](https://pandas.pydata.org/) and [matplotlib](https://matplotlib.org/) packages for data manipulation and visualisation, covered in the next blog post.

The basics of Python can be learnt relatively quickly. The aim is not to become a software engineer over-night. However, before we can run, we need to start walking, so it is important to spend some time learning the language of Python itself, before jumping into data analysis with Python.

Eventually, you will want to become a better programmer and at such point it's important to really hone your core coding skills and internalise this knowledge. If this is the stage you are at, I recommend some resources at the end of this post to enable to practice through coding challenges.

Before tackling this long read, I'll quickly summarise the contents of this entry. 

This blog goes into detail of how to do the following: 
*    Setup & Hello World
*    Variables & Data Types
*    Working With Strings
*    Working With Numbers
*    Getting Input From Users
*    Building a Basic Calculator
*    Mad Libs Game
*    Lists
*    List Functions
*    Tuples
*    Functions
*    Return Statement
*    If Statements
*    If Statements & Comparisons
*    Building a better Calculator
*    Dictionaries
*    While Loop
*    Building a Guessing Game
*    For Loops
*    Exponent Function
*    2D Lists & Nested Loops
*    Building a Translator
*    Comments
*    Try / Except
*    Reading Files
*    Writing to Files
*    Modules & Pip
*    Classes & Objects
*    Building a Multiple Choice Quiz
*    Object Functions
*    Inheritance
*    Python Interpreter

Finally, in the references at the bottom I've included some of my favourite tutorials and reference material for learning Pandas, including a handy cheat sheet.

Let's begin...

---

## 2. Getting Started
Before getting started, go to the toolbar and click 'Cell', followed by 'Run All' to run all the following commands.

---

## 3. Hello World

#### Theory
The first task any programmer does when learning a new language is learning how to print the phrase 'Hello World!', that is to display a sentence or 'string' on the screen. In this first section, we're not going to break from tradition and so we'll be doing just that (I promise, after this point, all the examples will be football related!).

To print a sentence, we call the `print()` function, as follows:

In [2]:
print('Hello World!')

Hello World!


Where the string in question, is included as an input (or argument) to the function in those parentheses.

#### Exercise
Print the name of your favourite football team.

In [ ]:
# WRITE YOUR CODE HERE

---

## 4. Comments

#### Theory
In the previous section, you were invited to to print the name of your favourite football team and in the cell, there was a comment. 

A comment in Python begins with the `#` symbol. In English, this is called the hash symbol (or octothorpe or pound character). 

In [8]:
# Print the number of goals scored
print(goals_scored)

8


Comments are very useful in programming because they explain what something does in English (or your chosen language). This is very important in professional programming when you will often have to pick up from where a previous employee left off and use their code (a fact of a programmer's life!)

They also develop part of a program if you need to remove them temporarily and do not want to delete what you just wrote.

#### Exercise
Print the name of your least favourite football team and write a comment above it to say in English that you are doing so.

In [ ]:
# WRITE YOUR CODE HERE

---

## 5. Variables

#### Theory
A variable can be created by the following:

In [6]:
goals_scored = 5

The variable `goals_scored` is created and a value of 5 is assigned to it using =. In Python, '=' is called the assignment operator.

The value assigned to the `goal_scored` variable can be displayed using the `print` function. Functions are called by putting parentheses after their name and putting inputs (or arguments) to the function in those parentheses.

In [4]:
print(goals_scored)

5


The value of a variable can be reassigned a new value using the assignmenter (=) operator, the same way in which the initial variable is created

In [7]:
goals_scored = goals_scored + 3
print(goals_scored)

8


In this example, we're conducting some basic arithmetic, adding '3' to the existing value of '5' to make 8.

#### Exercise
Create a variable for the goals conceded, `goals_conceded` and assign the value 10. In the following command then assign this a value 3 greater than the initial value.

In [ ]:
# WRITE THE FIRST PART OF YOUR CODE HERE

In [ ]:
# WRITE THE SECOND PART OF YOUR CODE HERE

---

## 6. Data Types

#### Theory
In programming, the data type is a very important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:


| Type Category     | Data Type                              |
|-------------------|----------------------------------------|
| Text              | `str`                                  |
| Numeric           | `int`, `float`, `complex`              |
| Sequence          | `list`, `tuple`, `range`               |
| Mapping           | `dict`                                 |
| Set               | `set`, `frozenset`                     |
| Boolean           | `bool`                                 |
| Binary            | `bytes`, `bytearray`, `memoryview`     |

The type of value assigned to a variable can be determined calling the `type` function, as follows:

In [9]:
type(goals_scored)

int

In this example, we can see the `goals_scored` variable created in the previous section has a `int` type, short for integer.

Some other examples of variables with different data types are the following:

In [13]:
# Assign value to variable
expected_goals = 1.7

# Print variable
print(expected_goals)

# Print variable type
type(expected_goals)

1.7


float

A `float` is a number with a decimal place - very useful for representing values such as Expected Goals, which are probabilities of a shot with a given state and conditions resulting in a goal.

In [11]:
# Assign value to variable
player_name = 'Roberto Larcos'

# Print variable
print(player_name)

# Print variable type
type(player_name)

Roberto Larcos


str

`str` is short for String.

Other data types including `bool` and then more complex data types such as `list`, `tuple`, `dict`, and `set` are covered in later sections.

![larcos](../../../img/roberto_larcos.jpeg)

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 7. Mathematical Operations

#### Theory
When working with numbers, a key thing is to be able to perform basic arithmetic. In section 5, we saw the use of the `+` operator for addition when reassigning a new value to the `goals_scored` variable. Python is also able to carry out the rest of the basic mathematical functions that you would find on your calculator:

| Operator       | Name                 | Description                                            |
|----------------|----------------------|--------------------------------------------------------|
| `a` + `b`      | Addition             | Sum of `a` and `b`                                     |
| `a` - `b`      | Subtraction          | Difference of a and `b`                                |
| `a` * `b`      | Multiplication       | Product of `a` and `b`                                 |
| `a` / `b`      | True division        | Quotient of `a` and `b`                                |
| `a` // `b`     | Floor division       | Quotient of `a` and `b`, removing fractional parts     |
| `a` % `b`      | Modulus	Integer     | remainder after division of `a` by `b`                 |
| `a` ** `b`     | Exponentiation       | a raised to the power of `b`                           |
| -`a`           | Negation             | The negative of `a`                                    |

One difference between Python and your calculator is that Python is able to do two kinds of division, not just one

The first, 'True division', is what you would expect on your calculator:

In [14]:
print(10 / 2)

5.0


In [15]:
print(9 / 2)

4.5


It always gives us a `float`.

The `//` operator gives us a result that's rounded down to the next integer.

In [16]:
print(10 // 2)

5


In [18]:
print(9 // 2)

4


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 8. Order of operations

#### Theory
When learning arithmetic in primary school, we learn the conventions about the order in which operations are evaluated. Examples of different mnemonics to remember these are:
*    **PEMDAS** - Parentheses, Exponents, Multiplication/Division, Addition/Subtractionl or
*    **BODMAS** - Brackets, Orders, Division/Multiplication, Addition/Subtraction.

Python also follows similar rules when performing calculations...

In [19]:
8 - 3 + 2

7

In [20]:
-3 + 4 * 2

5

Like mathematics, brackets can be used to force Python to evaluate subexpressions in whatever order you want, which can be observed in the following examples when determining the number of goals per game, `goals_per_game`.

In this example, integer values have been assigned to variables.

In [21]:
# Create variables for the number of goals scored and matches player in a season
goals_league = 15
goals_cups = 5
total_matches = 40

# Calculate the goals per game by dividing the sum total of goals by the number of matches
goals_per_game = goals_league + goals_cups / total_matches

# Print the variable for the calculated goals per game, goals_per_game
print(goals_per_game)

15.125


15.125 goals per game!? Now that would be a pretty unbelievable goal scorer! However, we know this is actually due to a mathematical error.

Let's try again with brackets...

In [22]:
# Create variables for the number of goals scored and matches player in a season
goals_league = 15
goals_cups = 5
total_matches = 40

# Calculate the goals per game by dividing the sum total of goals by the number of matches
goals_per_game = (goals_league + goals_cups) / total_matches

# Print the variable for the calculated goals per game, goals_per_game
print(goals_per_game)

0.5


We can see that our player is a 1 in 2 goal scorer.

#### Exercise
Calculate the 

In [ ]:
# WRITE YOUR CODE HERE

---

## 9. Functions for Numbers

#### Theory
We've currently seen the `print` and `type` functions, that display a value on screen and display an object type, respectively.

When working with numbers, there are a number of functions that are commonly used, that are quite self-explanatory, as you will see.

The first two, `min` and `max`, return the minimum and maximum of their arguments, respectively, as follows:

In [23]:
print(min(1, 2, 3))

1


In [24]:
print(max(1, 2, 3))

3


The can also be used when a numerical value is assigned to a variable, for example:

In [178]:
# Assign the number of goals scored to each player
naldorinho = 14
zinadine_ziderm = 11
gabriele_butatista = 18
ronarid = 20
david_backham = 7

# Determine what was the maximum number of goals scored
max(ronarid, zinadine_ziderm, naldorinho, gabriele_butatista, david_backham)

20

We can see that the player who scored the most number of goals was Ronarid with 20 goals.

![ronarid](../../../img/ronarid.jpeg)

`abs` returns the absolute value of an argument:

In [25]:
print(abs(32))

32


In [26]:
print(abs(-32))

32


`round` rounds a float to the nearest integer:

In [42]:
round(3.7)

4

Again, as demonstrated with the `max` function, the `abs` function can be called for a variable in which a numerical value as been assigned, for example:

In [175]:
# Assign the number of Expected Goals (xG) during a season assigned to the player
naldorinho = 16.7

# Determine what was the player's xG, rounded to the nearest significant figure
round(naldorinho)

17

![naldorinho](../../../img/naldorinho.jpeg)

#### Exercise
Create five variables, each one representing a goalkeeper and call the `min` function to determine which goalkeeper conceded the fewest goals.

For name inspirations, check out the [Unlicensed FC](https://www.instagram.com/unlicensed.fc/) Instagram page.

In [ ]:
# WRITE YOUR CODE HERE

---

## 10. Specifying a Variable Type

#### Theory
There are times when coding that you want to change the data type of a variable to another specified type. This can be done with **casting**. 

For the nerds - Python is an object-orientated language and because of this, it uses **classes** to define data types, including its primitive types (don't worry if you don't know what this means at this stage!).

Casting in Python can be done using the following:
*    `int()` - to integer;
*    `float()` - to float; and
*    `str()` - to string.

In the following subsections, the casting types have been split into **numerical** and **string** categories.

### Numerical Casting
In addition to being the names of Python's two main numerical types, `int` and `float` can also be called as functions, converting the the values between the parenthesis (the arguments) to the specified type, e.g.

In [179]:
print(float(7))

7.0


In [180]:
print(int(5.25))

5


The two numerical casting functions can even be called on strings, for example:

In [182]:
print(int('45') + 1)

46


In [184]:
print(float('93.20'))

93.2


As previously mentioned in section 5, we can check the type of the output by using the `type` function, e.g.

In [34]:
type(int(3.33))

int

### String Casting

In [186]:
str(9)

'9'

In [187]:
str(11.0)

'11.0'

Again, we can check the type of the output by using the `type` function, e.g.

In [188]:
type(str(9.0))

str

#### Exercise
Convert the following variables to the defined data type:
1.    A string type in integer;
2.    An integer type to float; and
3.    A float type to a string.

In [ ]:
# WRITE THE FIRST PART OF YOUR CODE HERE

In [ ]:
# WRITE THE SECOND PART OF YOUR CODE HERE

In [ ]:
# WRITE THE THIRD PART OF YOUR CODE HERE

---

## 11. Asking for Help

#### Theory
Up until this point, we have called a few functions in this notebook such as `print`, `type`, `abs`, `round`, `int`, `float`, and `string`.

Python has many, many built in functions and when you work with other libraries, you see a whole lot more! So what do we do if we've forgotten what a function does?

This is where the `help()` function comes in, for example:

In [43]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



In this example, `help()` is displaying two things:

1.    the header of that function `round(number, ndigits=None)`. For the `round()` function, help tells us that it can take an argument that we can describe as number. In additional, there is the option to provide a second argument for the number of digits (ndigits); and
2.    A brief descirption in English of what the function does.

**Note:** when looking up a function using the `help` function, make sure to pass in the name of the function itself and not the result of calling that function.

What happens if we invoke help on a call to the function round()?

In [39]:
help(round(-2.01))

Help on int object:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __ceil__(...)
 |      Ceiling of an Integral retur

Python evaluates an expression like this from the inside out. First it calculates the value of round(-2.01), then it provides help on the output of that expression.

(And it turns out to have a lot to say about integers! After we talk later about objects, methods, and attributes in Python, the help output above will make more sense.)

`round` is a very simple function with a short docstring. help shines even more when dealing with more complex, configurable functions like print. Don't worry if the following output looks inscrutable... for now, just see if you can pick anything new out from this help.

In [44]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



If you were looking for it, you might learn that print can take an argument called `sep`, and that this describes what we put between all the other arguments when we print them.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

In [ ]:
# AT THIS POINT THIS NOTEBOOK IS A COPY AND PASTE OF KAGGLE

---

## 10. Defining Your Own Custom Function

#### Theory
Builtin functions are great, but we can only get so far with them before we need to start defining our own functions. Below is a simple example.

In [45]:
def least_difference(a, b, c):
    diff1 = abs(a - b)
    diff2 = abs(b - c)
    diff3 = abs(a - c)
    return min(diff1, diff2, diff3)

This creates a function called `least_difference`, which takes three arguments, `a`, `b`, and `c`.

Functions start with a header introduced by the `def` keyword. The indented block of code following the : is run when the function is called.

`return` is another keyword uniquely associated with functions. When Python encounters a return statement, it exits the function immediately, and passes the value on the right hand side to the calling context.

Is it clear what `least_difference()` does from the source code? If we're not sure, we can always try it out on a few examples:

In [46]:
least_difference(1, 10, 100)

9

In [47]:
least_difference(1, 10, 10)

0

In [48]:
least_difference(5, 6, 7)

1

Or maybe the help() function can tell us something about it.

In [49]:
help(least_difference)

Help on function least_difference in module __main__:

least_difference(a, b, c)



Python isn't smart enough to read my code and turn it into a nice English description. However, when I write a function, I can provide a description in what's called the **docstring**.

In [51]:
def least_difference(a, b, c):
    '''Return the smallest difference between any two numbers
    among a, b and c.
    
    >>> least_difference(1, 5, -5)
    4
    '''
    diff1 = abs(a - b)
    diff2 = abs(b - c)
    diff3 = abs(a - c)
    return min(diff1, diff2, diff3)

The docstring is a triple-quoted string (which may span multiple lines) that comes immediately after the header of a function. When we call `help()` on a function, it shows the docstring.

In [52]:
help(least_difference)

Help on function least_difference in module __main__:

least_difference(a, b, c)
    Return the smallest difference between any two numbers
    among a, b and c.
    
    >>> least_difference(1, 5, -5)
    4



The last two lines of the docstring are an example function call and result. (The `>>>` is a reference to the command prompt used in Python interactive shells.) Python doesn't run the example call - it's just there for the benefit of the reader. The convention of including 1 or more example calls in a function's docstring is far from universally observed, but it can be very effective at helping someone understand your function. For a real-world example, see [this docstring for the numpy function](https://github.com/numpy/numpy/blob/v1.14.2/numpy/lib/twodim_base.py#L140-L194) `np.eye`.

Good programmers use docstrings unless they expect to throw away the code soon after it's used (which is rare). So, you should start writing docstrings, too!

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 11. Functions Don't Always Return a Value

#### Theory
What would happen if we didn't include the return keyword in our function?

In [53]:
# Define a function for 'least_difference' i.e. the smallest difference between any two numbers among a, b and c
def least_difference(a, b, c):
    """Return the smallest difference between any two numbers
    among a, b and c.
    """
    diff1 = abs(a - b)
    diff2 = abs(b - c)
    diff3 = abs(a - c)
    min(diff1, diff2, diff3)
    
# Call 'least_difference' function
least_difference(1, 10, 100)

No output.

Python allows us to define such functions. The result of calling them is the special value `None`. (This is similar to the concept of "null" in other languages.)

Without a `return` statement, `least_difference` is completely pointless, but a function with side effects may do something useful without returning anything. We've already seen two examples of this: `print()` and `help()` don't return anything. We only call them for their side effects (putting some text on the screen). Other examples of useful side effects include writing to a file, or modifying an input.

---

## 12. Default Arguments of Functions

#### Theory
When we call `help(print)`, we saw that the print function has several optional arguments. For example, we can specify a value for sep to put some special string in between our printed arguments:

In [54]:
print(1, 2, 3, sep=' < ')

1 < 2 < 3


But if we don't specify a value, sep is treated as having a default value of ' ' (a single space).

In [55]:
print(1, 2, 3)

1 2 3


Adding optional arguments with default values to the functions we define turns out to be pretty easy:

In [56]:
def greet(who="Colin"):
    print("Hello,", who)
    
greet()
greet(who="Kaggle")
# (In this case, we don't need to specify the name of the argument, because it's unambiguous.)
greet("world")

Hello, Colin
Hello, Kaggle
Hello, world


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 13. Functions Applied to Functions

#### Theory
Here's something that's powerful, though it can feel very abstract at first. You can supply functions as arguments to other functions. Some example may make this clearer:

In [57]:
def mult_by_five(x):
    return 5 * x

def call(fn, arg):
    """Call fn on arg"""
    return fn(arg)

def squared_call(fn, arg):
    """Call fn on the result of calling fn on arg"""
    return fn(fn(arg))

print(call(mult_by_five, 1),
      squared_call(mult_by_five, 1), 
      sep='\n', # '\n' is the newline character - it starts a new line
     )

5
25


Functions that operate on other functions are called "higher-order functions." You probably won't write your own for a little while. But there are higher-order functions built into Python that you might find useful to call.

Here's an interesting example using the `max` function.

By default, `max` returns the largest of its arguments. But if we pass in a function using the optional key argument, it returns the argument `x` that maximizes `key(x)` (aka the 'argmax').

In [58]:
def mod_5(x):
    """Return the remainder of x after dividing by 5"""
    return x % 5

print('Which number is biggest?',
      max(100, 51, 14),
      'Which number is the biggest modulo 5?',
      max(100, 51, 14, key=mod_5),
      sep='\n',
    )

Which number is biggest?
100
Which number is the biggest modulo 5?
14


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 14. Booleans
We previously saw in section 5 a list of all the data types in Python, one of which is the Boolean data type - `bool`. It has two possible values: `True` and `False`.

In [59]:
x = True
print(x)
print(type(x))

True
<class 'bool'>


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 15. Comparison Operations

#### Theory
Rather than putting `True` or `False` directly in our code, we usually get boolean values from **boolean operators**. These are operators that answer yes/no questions. We'll go through some of these operators below.

#### Comparison Operations:

| Operator        | Name                                 |
|-----------------|--------------------------------------|
| `a` == `b`      | `a` equal to `b`                     |
| `a` == `b`      | `a` not qual to `b`                  |
| `a` > `b`       | `a` less than `b`                    |
| `a` > `b`       | `a` greater than `b`                 |
| `a` <= `b`      | `a` less than or equal to `b`        |
| `a` <= `b`      | `a` greater than or equal to `b`     |

In [60]:
def can_run_for_president(age):
    """Can someone of the given age run for president in the US?"""
    # The US Constitution says you must be at least 35 years old
    return age >= 35

print("Can a 19-year-old run for president?", can_run_for_president(19))
print("Can a 45-year-old run for president?", can_run_for_president(45))

Can a 19-year-old run for president? False
Can a 45-year-old run for president? True


Comparisons frequently work like you'd hope

In [61]:
3.0 == 3

True

But sometimes they can be tricky

In [62]:
'3' == 3

False

Comparison operators can be combined with the arithmetic operators we've already seen to express a virtually limitless range of mathematical tests. For example, we can check if a number is odd by checking that the modulus with 2 returns 1:

In [63]:
def is_odd(n):
    return (n % 2) == 1

print("Is 100 odd?", is_odd(100))
print("Is -1 odd?", is_odd(-1))

Is 100 odd? False
Is -1 odd? True


Remember to use `==` instead of `=` when making comparisons. If you write `n == 2` you are asking about the value of n. When you write `n = 2` you are changing the value of n.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 16. Combining Boolean Values

#### Theory
You can combine boolean values using the standard concepts of "and", "or", and "not". In fact, the words to do this are: `and`, `or`, and `not`.

With these, we can make our `can_run_for_president` function more accurate.

In [64]:
def can_run_for_president(age, is_natural_born_citizen):
    """Can someone of the given age and citizenship status run for president in the US?"""
    # The US Constitution says you must be a natural born citizen *and* at least 35 years old
    return is_natural_born_citizen and (age >= 35)

print(can_run_for_president(19, True))
print(can_run_for_president(55, False))
print(can_run_for_president(55, True))

False
False
True


Quick, can you guess the value of this expression?

In [65]:
True or True and False

True

To answer this, you'd need to figure out the order of operations.

For example, `and` is evaluated before or. That's why the first expression above is `True`. If we evaluated it from left to right, we would have calculated `True or True` first (which is `True`), and then taken the and of that result with `False`, giving a final value of `False`.

You could try to memorise the [order of precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence), but a safer bet is to just use liberal parentheses. Not only does this help prevent bugs, it makes your intentions clearer to anyone who reads your code.

For example, consider the following expression:

In [66]:
prepared_for_weather = have_umbrella or rain_level < 5 and have_hood or not rain_level > 0 and is_workday

NameError: name 'have_umbrella' is not defined

I'm trying to say that I'm safe from today's weather....

*    if I have an umbrella...
*    or if the rain isn't too heavy and I have a hood...
*    otherwise, I'm still fine unless it's raining and it's a workday

But not only is my Python code hard to read, it has a bug. We can address both problems by adding some parentheses:

In [ ]:
prepared_for_weather = have_umbrella or (rain_level < 5 and have_hood) or not (rain_level > 0 and is_workday)

You can add even more parentheses if you think it helps readability:

In [ ]:
prepared_for_weather = have_umbrella or ((rain_level < 5) and have_hood) or (not (rain_level > 0 and is_workday))

We can also split it over multiple lines to emphasize the 3-part structure described above:

In [ ]:
prepared_for_weather = (have_umbrella 
                        or ((rain_level < 5) and have_hood) 
                        or (not (rain_level > 0 and is_workday))
                       )

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 17. Conditionals

#### Theory
Booleans are most useful when combined with conditional statements, using the keywords `if`, `elif`, and `else`.

Conditional statements, often referred to as **if-then** statements, let you control what pieces of code are run based on the value of some Boolean condition. Here's an example:

In [67]:
def inspect(x):
    if x == 0:
        print(x, "is zero")
    elif x > 0:
        print(x, "is positive")
    elif x < 0:
        print(x, "is negative")
    else:
        print(x, "is unlike anything I've ever seen...")

inspect(0)
inspect(-15)

0 is zero
-15 is negative


The `if` and `else` keywords are often used in other languages; its more unique keyword is `elif`, a contraction of "else if". In these conditional clauses, `elif` and `else` blocks are optional; additionally, you can include as many `elif` statements as you would like.

Note especially the use of colons (`:`) and whitespace to denote separate blocks of code. This is similar to what happens when we define a function - the function header ends with `:`, and the following line is indented with four spaces. All subsequent indented lines belong to the body of the function, until we encounter an unindented line, ending the function definition.

In [68]:
def f(x):
    if x > 0:
        print("Only printed when x is positive; x =", x)
        print("Also only printed when x is positive; x =", x)
    print("Always printed, regardless of x's value; x =", x)

f(1)
f(0)

Only printed when x is positive; x = 1
Also only printed when x is positive; x = 1
Always printed, regardless of x's value; x = 1
Always printed, regardless of x's value; x = 0


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 18. Boolean conversion

#### Theory
We've seen `int()`, which turns things into ints, and `float()`, which turns things into floats, so you might not be surprised to hear that Python has a `bool()` function which turns things into bools.

In [69]:
print(bool(1)) # all numbers are treated as true, except 0
print(bool(0))
print(bool("asf")) # all strings are treated as true, except the empty string ""
print(bool(""))
# Generally empty sequences (strings, lists, and other types we've yet to see like lists and tuples)
# are "falsey" and the rest are "truthy"

True
False
True
False


We can use non-boolean objects in `if` conditions and other places where a boolean would be expected. Python will implicitly treat them as their corresponding boolean value:

In [70]:
if 0:
    print(0)
elif "spam":
    print("spam")

spam


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 19. Lists

#### Theory
We previously saw in section 6 a list of all the data types in Python, one of which is the List data type - `list`.

Lists in Python represent ordered sequences of values. 

The following is an example of a list with numerical values, each one representing the number of goals Lionel Messi has scored in LaLiga for each season between 04/05 and 20/21 [[link](https://www.transfermarkt.com/lionel-messi/leistungsdatendetails/spieler/28003/plus/0?saison=&verein=&liga=&wettbewerb=ES1&pos=&trainer_id=)]:

In [190]:
# Create a list of integers
messi_goals_per_season = [1, 6, 14, 10, 23, 34, 31, 50, 46, 28, 43, 26, 37, 34, 36, 25, 30]

We can put other types of things in lists:

In [194]:
# Create a list of strings
french_xi = ['Barllez', 'Desarie', 'Bulan', 'Rezarzu', 'Therum', 'Delsham', 'Patit', 'Ziderm', 'Djolkef', 'Henlie', 'Guivaus']

# Print the list
print(french_xi)

['Barllez', 'Desarie', 'Bulan', 'Rezarzu', 'Therum', 'Delsham', 'Patit', 'Ziderm', 'Djolkef', 'Henlie', 'Guivaus']


![brazil_france](../../../img/brazil_france_iss_98.jpeg)

Lists can also be made from other lists.

In the following example, a list of players is assigned to a variable, `brazil_goalkeeper`, `brazil defenders`, `brazil_midfielders`, and `brazil_forwards`. A list is created from these defined lists to create the variable `brazil_xi`.

In [197]:
# Create lists of strings
brazil_goalkeeper = ['Taffares']
brazil_defenders = ['Oudier', 'Bariano', 'R. Carls', 'Kafou']
brazil_midfielders = ['Donka', 'Rivalno', 'Sunpaio', 'Rioneld']
brazil_forwards = ['Berett', 'Ronarid']

# Create a list of lists
brazil_xi = [brazil_goalkeeper, brazil_defenders, brazil_midfielders, brazil_forwards]              

# Print the list (of lists)
print(brazil_xi)

[['Taffares'], ['Oudier', 'Bariano', 'R. Carls', 'Kafou'], ['Donka', 'Rivalno', 'Sunpaio', 'Rioneld'], ['Berett', 'Ronarid']]


Lists can also contain a mix of different data types of variables.

The following list is made up of bio information for Kylian Mbappé [[link](https://www.transfermarkt.com/kylian-mbappe/profil/spieler/342229)].

In [199]:
# Create lists of variables with different data types
mbappe_bio = ['Kylian Mbappé Lottin', 22, 1.78, 'French', 'Right Footed']

# Print the list
print(mbappe_bio)

['Kylian Mbappé Lottin', 22, 1.78, 'French', 'Right Footed']


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 20. Indexing

#### Theory
You can access individual parts of a list (its elements) with square brackets.

The following code creates a list of Italian players. The subsequent command print the first element of the list, which in this example, is the goalkeeper in the starting XI.

**Note:** Python uses zero-based indexing i.e. it starts counting from 0, not 1, so the first element of the list has index 0.

In [201]:
# Create a list of strings
italy_xi = ['Peglioca',     # each string can be entered on a different line for easy readability
            'Costacuss',
            'Cannavar',
            'Nista',
            'Di Bialio',
            'Albertoni',
            'Malodini',
            'Di Rebio',
            'D. Faggio',
            'Del Paolo',
            'Vieni'
           ]

# Print the list
print(italy_xi)

['Peglioca', 'Costacuss', 'Cannavar', 'Nista', 'Di Bialio', 'Albertoni', 'Malodini', 'Di Rebio', 'D. Faggio', 'Del Paolo', 'Vieni']


In [202]:
italy_xi[0]

'Peglioca'

Who is the next player on the team sheet i.e. the second element in the list:

In [204]:
italy_xi[1]

'Costacuss'

![brazil_france](../../../img/italy_germany_iss_98.jpeg)

Which player is the last player on the team sheet i.e. the last element in the list?

The last element in a list can be access with negative numbers, starting from -1:

In [205]:
italy_xi[-1]

'Vieni'

In [206]:
italy_xi[-2]

'Del Paolo'

#### Exercise
Create a list of the German XI, from the image above from which take the following indexes:
*    the goalkeeper of the starting XI i.e. the first element of the list;
*    the last striker of the starting XI i.e. the last element of the list; and
*    the last four players of the starting XI i.e. the last four elements of the list.

In [ ]:
# WRITE THE FIRST PART OF YOUR CODE HERE

In [ ]:
# WRITE THE SECOND PART OF YOUR CODE HERE

In [ ]:
# WRITE THE THIRD PART OF YOUR CODE HERE

In [ ]:
# WRITE THE FORTH PART OF YOUR CODE HERE

---

## 21. Slicing

#### Theory
Who are the first five players on the team sheet? I.e. the first five elements in the list. This can be answerwed using slicing:

In [207]:
italy_xi[0:5]

['Peglioca', 'Costacuss', 'Cannavar', 'Nista', 'Di Bialio']

`italy_xi[0:5]` is our way of asking for the elements of `italy_xi` starting from index 0 and continuing up to but not including index 5.

The starting and ending indices are both optional. If we leave out the start index, it's assumed to be 0. So I could rewrite the expression above as:

In [208]:
italy_xi[:5]

['Peglioca', 'Costacuss', 'Cannavar', 'Nista', 'Di Bialio']

If I leave out the end index, it's assumed to be the length of the list.

In [210]:
italy_xi[5:]

['Albertoni', 'Malodini', 'Di Rebio', 'D. Faggio', 'Del Paolo', 'Vieni']

i.e. the expression above means "give me all the players on the team sheet from index 5 onwards".

We can also use negative indices when slicing:

In [211]:
# All the players on the team sheet except the first and last
italy_xi[1:-1]

['Costacuss',
 'Cannavar',
 'Nista',
 'Di Bialio',
 'Albertoni',
 'Malodini',
 'Di Rebio',
 'D. Faggio',
 'Del Paolo']

In [212]:
# The last 2 players i.e. just the forwards
italy_xi[-2:]

['Del Paolo', 'Vieni']

#### Exercise
Using the German XI list created in the previous exercise, slice the list as follows:
*    
*    
*    
*    

In [214]:
# AT THIS POINT THIS NOTEBOOK IS A COPY AND PASTE OF KAGGLE

---

## 22. Changing lists

#### Theory
Lists are "mutable", meaning they can be modified "in place".

One way to modify a list is to assign to an index or slice expression.

For example, let's say we want to rename Mars:

In [83]:
planets[3] = 'Malacandra'
planets

['Mercury',
 'Venus',
 'Earth',
 'Malacandra',
 'Jupiter',
 'Saturn',
 'Uranus',
 'Neptune']

Hm, that's quite a mouthful. Let's compensate by shortening the names of the first 3 planets.

In [84]:
planets[:3] = ['Mur', 'Vee', 'Ur']
print(planets)
# That was silly. Let's give them back their old names
planets[:4] = ['Mercury', 'Venus', 'Earth', 'Mars',]

['Mur', 'Vee', 'Ur', 'Malacandra', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 23. List functions

#### Theory
Python has several useful functions for working with lists.

`len` gives the length of a list:

In [85]:
# How many planets are there?
len(planets)

8

`sorted` returns a sorted version of a list:

In [86]:
# The planets sorted in alphabetical order
sorted(planets)

['Earth', 'Jupiter', 'Mars', 'Mercury', 'Neptune', 'Saturn', 'Uranus', 'Venus']

`sum` does what you might expect:

In [87]:
primes = [2, 3, 5, 7]
sum(primes)

17

We've previously used the `min` and `max` to get the minimum or maximum of several arguments. But we can also pass in a single list argument.

In [88]:
max(primes)

7

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 24. Objects

#### Theory
I've used the term 'object' a lot so far - you may have even read that everything in Python is an object. What does that mean?

In short, objects carry some things around with them. You access that stuff using Python's dot syntax.

For example, numbers in Python carry around an associated variable called `imag` representing their imaginary part. (You'll probably never need to use this unless you're doing some very weird math.)

In [89]:
x = 12
# x is a real number, so its imaginary part is 0.
print(x.imag)
# Here's how to make a complex number, in case you've ever been curious:
c = 12 + 3j
print(c.imag)

0
3.0


The things an object carries around can also include functions. A function attached to an object is called a method. (Non-function things attached to an object, such as `imag`, are called attributes).

For example, numbers have a method called `bit_length`. Again, we access it using dot syntax:

In [90]:
x.bit_length

<function int.bit_length()>

To actually call it, we add parentheses:

In [91]:
x.bit_length()

4

In [92]:
help(x.bit_length)

Help on built-in function bit_length:

bit_length() method of builtins.int instance
    Number of bits necessary to represent self in binary.
    
    >>> bin(37)
    '0b100101'
    >>> (37).bit_length()
    6



The examples above were utterly obscure. None of the types of objects we've looked at so far (numbers, functions, booleans) have attributes or methods you're likely ever to use.

But it turns out that lists have several methods which you'll use all the time.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 25. List methods

#### Theory
`list.append` modifies a list by adding an item to the end:

In [93]:
# Pluto is a planet darn it!
planets.append('Pluto')

Why does the cell above have no output? Let's check the documentation by calling `help(planets.append)`.

`append` is a method carried around by all objects of type list, not just `planets`, so we also could have called `help(list.append)`. However, if we try to call `help(append)`, Python will complain that no variable exists called "append". The "append" name only exists within lists - it doesn't exist as a standalone name like builtin functions such as `max` or `len`.

In [94]:
help(planets.append)

Help on built-in function append:

append(object, /) method of builtins.list instance
    Append object to the end of the list.



The `-> None` part is telling us that list.append doesn't return anything. But if we check the value of planets, we can see that the method call modified the value of planets:

In [95]:
planets

['Mercury',
 'Venus',
 'Earth',
 'Mars',
 'Jupiter',
 'Saturn',
 'Uranus',
 'Neptune',
 'Pluto']

`list.pop` removes and returns the last element of a list:

In [96]:
planets.pop()

'Pluto'

In [97]:
planets

['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']

It's possible also search lists. Where does Earth fall in the order of planets? We can get its index using the `list.index` method.

In [98]:
planets.index('Earth')

2

It comes third (i.e. at index 2 - 0 indexing!).

At what index does Pluto occur?

In [99]:
planets.index('Pluto')

ValueError: 'Pluto' is not in list

Oh, that's right...

To avoid unpleasant surprises like this, we can use the in operator to determine whether a list contains a particular value:

In [100]:
# Is Earth a planet?
"Earth" in planets

True

In [101]:
# Is Calbefraques a planet?
"Calbefraques" in planets

False

There are a few more interesting list methods we haven't covered. If you want to learn about all the methods and attributes attached to a particular object, we can call `help()` on the object itself. For example, `help(planets)` will tell us about all the list methods:

In [102]:
help(planets)

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate sign

Lists have lots of methods with weird-looking names like `__eq__` and `__iadd__`. Don't worry too much about these for now. (You'll probably never call such methods directly. But they get called behind the scenes when we use syntax like indexing or comparison operators.) The most interesting methods are toward the bottom of the list (`append`, `clear`, `copy`, etc.).

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 26. Tuples

#### Theory
Tuples are almost exactly the same as lists. They differ in just two ways.

1: The syntax for creating them uses parentheses instead of square brackets

In [103]:
t = (1, 2, 3)

In [104]:
t = 1, 2, 3 # equivalent to above
t

(1, 2, 3)

2: They cannot be modified (they are immutable).

In [105]:
t[0] = 100

TypeError: 'tuple' object does not support item assignment

Tuples are often used for functions that have multiple return values.

For example, the `as_integer_ratio()` method of float objects returns a numerator and a denominator in the form of a tuple:

In [106]:
x = 0.125
x.as_integer_ratio()

(1, 8)

These multiple return values can be individually assigned as follows:

In [107]:
numerator, denominator = x.as_integer_ratio()
print(numerator / denominator)

0.125


Finally we have some insight into the classic Stupid Python Trick™ for swapping two variables!

In [108]:
a = 1
b = 0
a, b = b, a
print(a, b)

0 1


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 27. Loops

#### Theory
Loops are a way to repeatedly execute some code. Here's an example:

In [109]:
planets = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']

for planet in planets:
    print(planet, end=' ') # print all on same line

Mercury Venus Earth Mars Jupiter Saturn Uranus Neptune 

The `for` loop specifies

*    the variable name to use (in this case, planet)
*    the set of values to loop over (in this case, planets)

You use the word "`in`" to link them together.

The object to the right of the "`in`" can be any object that supports iteration. Basically, if it can be thought of as a group of things, you can probably loop over it. In addition to lists, we can iterate over the elements of a tuple:

In [110]:
multiplicands = (2, 2, 2, 3, 3, 5)
product = 1
for mult in multiplicands:
    product = product * mult
product

360

You can even loop through each character in a string:

In [111]:
s = 'steganograpHy is the practicE of conceaLing a file, message, image, or video within another fiLe, message, image, Or video.'
msg = ''
# print all the uppercase letters in s, one at a time
for char in s:
    if char.isupper():
        print(char, end='')  

HELLO

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 28. Range

#### Theory
`range()` is a function that returns a sequence of numbers. It turns out to be very useful for writing loops.

For example, if we want to repeat some action 5 times:

In [112]:
for i in range(5):
    print("Doing important work. i =", i)

Doing important work. i = 0
Doing important work. i = 1
Doing important work. i = 2
Doing important work. i = 3
Doing important work. i = 4


#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 29. `while` loops

#### Theory
The other type of loop in Python is a while loop, which iterates until some condition is met:

In [113]:
i = 0
while i < 10:
    print(i, end=' ')
    i += 1 # increase the value of i by 1

0 1 2 3 4 5 6 7 8 9 

The argument of the `while` loop is evaluated as a boolean statement, and the loop is executed until the statement evaluates to False.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 30. List comprehensions

#### Theory
List comprehensions are one of Python's most beloved and unique features. The easiest way to understand them is probably to just look at a few examples:

In [114]:
squares = [n**2 for n in range(10)]
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

Here's how we would do the same thing without a list comprehension:

In [115]:
squares = []
for n in range(10):
    squares.append(n**2)
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

We can also add an `if` condition:

In [116]:
short_planets = [planet for planet in planets if len(planet) < 6]
short_planets

['Venus', 'Earth', 'Mars']

(If you're familiar with SQL, you might think of this as being like a "WHERE" clause)

Here's an example of filtering with an if condition and applying some transformation to the loop variable:

In [117]:
# str.upper() returns an all-caps version of a string
loud_short_planets = [planet.upper() + '!' for planet in planets if len(planet) < 6]
loud_short_planets

['VENUS!', 'EARTH!', 'MARS!']

People usually write these on a single line, but you might find the structure clearer when it's split up over 3 lines:

In [118]:
[
    planet.upper() + '!' 
    for planet in planets 
    if len(planet) < 6
]

['VENUS!', 'EARTH!', 'MARS!']

(Continuing the SQL analogy, you could think of these three lines as SELECT, FROM, and WHERE)

The expression on the left doesn't technically have to involve the loop variable (though it'd be pretty unusual for it not to). What do you think the expression below will evaluate to? Press the 'output' button to check.

In [119]:
[32 for planet in planets]

[32, 32, 32, 32, 32, 32, 32, 32]

List comprehensions combined with functions like min, max, and sum can lead to impressive one-line solutions for problems that would otherwise require several lines of code.

For example, compare the following two cells of code that do the same thing.

In [120]:
def count_negatives(nums):
    """Return the number of negative numbers in the given list.
    
    >>> count_negatives([5, -1, -2, 0, 3])
    2
    """
    n_negative = 0
    for num in nums:
        if num < 0:
            n_negative = n_negative + 1
    return n_negative

Here's a solution using a list comprehension:

In [121]:
def count_negatives(nums):
    return len([num for num in nums if num < 0])

Much better, right?

Well if all we care about is minimizing the length of our code, this third solution is better still!

In [122]:
def count_negatives(nums):
    # Reminder: in the "booleans and conditionals" exercises, we learned about a quirk of 
    # Python where it calculates something like True + True + False + True to be equal to 3.
    return sum([num < 0 for num in nums])

Which of these solutions is the "best" is entirely subjective. Solving a problem with less code is always nice, but it's worth keeping in mind the following lines from [The Zen of Python](https://en.wikipedia.org/wiki/Zen_of_Python):

> Readability counts.
> Explicit is better than implicit.

So, use these tools to make compact readable programs. But when you have to choose, favor code that is easy for others to understand.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 31. Strings and their Syntax

#### Theory
One place where the Python language really shines is in the manipulation of strings. This section will cover some of Python's built-in string methods and formatting operations.

Such string manipulation patterns come up often in the context of data science work.

You've already seen plenty of strings in examples during the previous lessons, but just to recap, strings in Python can be defined using either single or double quotations. They are functionally equivalent.

In [123]:
x = 'Pluto is a planet'
y = "Pluto is a planet"
x == y

True

Double quotes are convenient if your string contains a single quote character (e.g. representing an apostrophe).

Similarly, it's easy to create a string that contains double-quotes if you wrap it in single quotes:

In [124]:
print("Pluto's a planet!")
print('My dog is named "Pluto"')

Pluto's a planet!
My dog is named "Pluto"


If we try to put a single quote character inside a single-quoted string, Python gets confused:

In [125]:
'Pluto's a planet!'

SyntaxError: invalid syntax (<ipython-input-125-a43631749f52>, line 1)

We can fix this by "escaping" the single quote with a backslash.

In [126]:
'Pluto\'s a planet!'

"Pluto's a planet!"

The table below summarizes some important uses of the backslash character.

| What you type...       | What you get     | example                    | print(example)              |
|------------------------|------------------|----------------------------|-----------------------------|
| \\'                     | '                | 'What\\'s up?'              | What's up?                |
| \\"                     | "                | "That's \\"cool\\""          | That's "cool"            |
| \\\                     | \                | "Look, a mountain: /\\\"     | Look, a mountain: /\     |
| \\n                     |                  | "1\\n2 3"                   | 1<br>2 3                  |

The last sequence, `\n`, represents the newline character. It causes Python to start a new line.

In [139]:
hello = "hello\nworld"
print(hello)

hello
world


In addition, Python's triple quote syntax for strings lets us include newlines literally (i.e. by just hitting 'Enter' on our keyboard, rather than using the special '\n' sequence). We've already seen this in the docstrings we use to document our functions, but we can use them anywhere we want to define a string.

In [140]:
triplequoted_hello = """hello
world"""
print(triplequoted_hello)
triplequoted_hello == hello

hello
world


True

The `print()` function automatically adds a newline character unless we specify a value for the keyword argument end other than the default value of `'\n'`:

In [141]:
print("hello")
print("world")
print("hello", end='')
print("pluto", end='')

hello
world
hellopluto

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 32. Strings are sequences

#### Theory
Strings can be thought of as sequences of characters. Almost everything we've seen that we can do to a list, we can also do to a string.

In [142]:
# Indexing
planet = 'Pluto'
planet[0]

'P'

In [143]:
# Slicing
planet[-3:]

'uto'

In [144]:
# How long is this string?
len(planet)

5

In [145]:
# Yes, we can even loop over them
[char+'! ' for char in planet]

['P! ', 'l! ', 'u! ', 't! ', 'o! ']

But a major way in which they differ from lists is that they are immutable. We can't modify them.

In [146]:
planet[0] = 'B'
# planet.append doesn't work either

TypeError: 'str' object does not support item assignment

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 33. String methods

#### Theory
Like `list`, the type `str` has lots of very useful methods. I'll show just a few examples here.

In [147]:
# ALL CAPS
claim = "Pluto is a planet!"
claim.upper()

'PLUTO IS A PLANET!'

In [148]:
# all lowercase
claim.lower()

'pluto is a planet!'

In [149]:
# Searching for the first index of a substring
claim.index('plan')

11

In [150]:
claim.startswith(planet)

True

In [151]:
claim.endswith('dwarf planet')

False

#### Exercise
...

---

## 33. Going between strings and lists: `.split()` and `.join()`

#### Theory
`str.split()` turns a string into a list of smaller strings, breaking on whitespace by default. This is super useful for taking you from one big string to a list of words.

In [152]:
words = claim.split()
words

['Pluto', 'is', 'a', 'planet!']

Occasionally you'll want to split on something other than whitespace:

In [153]:
datestr = '1956-01-31'
year, month, day = datestr.split('-')

`str.join()` takes us in the other direction, sewing a list of strings up into one long string, using the string it was called on as a separator.

In [154]:
'/'.join([month, day, year])

'01/31/1956'

In [155]:
# Yes, we can put unicode characters right in our string literals :)
' 👏 '.join([word.upper() for word in words])

'PLUTO 👏 IS 👏 A 👏 PLANET!'

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 34. Building strings with `.format()`

#### Theory
Python lets us concatenate strings with the + operator.

In [156]:
planet + ', we miss you.'

'Pluto, we miss you.'

If we want to throw in any non-string objects, we have to be careful to call `str()` on them first

In [157]:
position = 9
planet + ", you'll always be the " + position + "th planet to me."

TypeError: can only concatenate str (not "int") to str

In [158]:
planet + ", you'll always be the " + str(position) + "th planet to me."

"Pluto, you'll always be the 9th planet to me."

This is getting hard to read and annoying to type. `str.format()` to the rescue.

In [159]:
"{}, you'll always be the {}th planet to me.".format(planet, position)

"Pluto, you'll always be the 9th planet to me."

So much cleaner! We call `.format()` on a "format string", where the Python values we want to insert are represented with `{}` placeholders.

Notice how we didn't even have to call `str()` to convert `position` from an int. `format()` takes care of that for us.

If that was all that `format()` did, it would still be incredibly useful. But as it turns out, it can do a lot more. Here's just a taste:

In [160]:
pluto_mass = 1.303 * 10**22
earth_mass = 5.9722 * 10**24
population = 52910390
#         2 decimal points   3 decimal points, format as percent     separate with commas
"{} weighs about {:.2} kilograms ({:.3%} of Earth's mass). It is home to {:,} Plutonians.".format(
    planet, pluto_mass, pluto_mass / earth_mass, population,
)

"Pluto weighs about 1.3e+22 kilograms (0.218% of Earth's mass). It is home to 52,910,390 Plutonians."

In [161]:
# Referring to format() arguments by index, starting from 0
s = """Pluto's a {0}.
No, it's a {1}.
{0}!
{1}!""".format('planet', 'dwarf planet')
print(s)

Pluto's a planet.
No, it's a dwarf planet.
planet!
dwarf planet!


You could probably write a short book just on str.format, so I'll stop here, and point you to [pyformat.info](https://pyformat.info/) and [the official docs](https://docs.python.org/3/library/string.html#formatstrings) for further reading.

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 35. Dictionaries

#### Theory
Dictionaries are a built-in Python data structure for mapping keys to values.

In [162]:
numbers = {'one':1, 'two':2, 'three':3}

In this case `'one'`, `'two'`, and `'three'` are the **keys**, and 1, 2 and 3 are their corresponding values.

Values are accessed via square bracket syntax similar to indexing into lists and strings.

In [163]:
numbers['one']

1

We can use the same syntax to add another key, value pair

In [164]:
numbers['eleven'] = 11
numbers

{'one': 1, 'two': 2, 'three': 3, 'eleven': 11}

Or to change the value associated with an existing key

In [165]:
numbers['one'] = 'Pluto'
numbers

{'one': 'Pluto', 'two': 2, 'three': 3, 'eleven': 11}

Python has dictionary comprehensions with a syntax similar to the list comprehensions we saw in the previous tutorial.

In [166]:
planets = ['Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune']
planet_to_initial = {planet: planet[0] for planet in planets}
planet_to_initial

{'Mercury': 'M',
 'Venus': 'V',
 'Earth': 'E',
 'Mars': 'M',
 'Jupiter': 'J',
 'Saturn': 'S',
 'Uranus': 'U',
 'Neptune': 'N'}

The `in` operator tells us whether something is a key in the dictionary

In [167]:
'Saturn' in planet_to_initial

True

In [168]:
'Betelgeuse' in planet_to_initial

False

A for loop over a dictionary will loop over its keys

In [169]:
for k in numbers:
    print("{} = {}".format(k, numbers[k]))

one = Pluto
two = 2
three = 3
eleven = 11


We can access a collection of all the keys or all the values with `dict.keys()` and `dict.values()`, respectively.

In [170]:
# Get all the initials, sort them alphabetically, and put them in a space-separated string.
' '.join(sorted(planet_to_initial.values()))

'E J M M N S U V'

The very useful `dict.items()` method lets us iterate over the keys and values of a dictionary simultaneously. (In Python jargon, an **item** refers to a key, value pair)

In [171]:
for planet, initial in planet_to_initial.items():
    print("{} begins with \"{}\"".format(planet.rjust(10), initial))

   Mercury begins with "M"
     Venus begins with "V"
     Earth begins with "E"
      Mars begins with "M"
   Jupiter begins with "J"
    Saturn begins with "S"
    Uranus begins with "U"
   Neptune begins with "N"


To read a full inventory of dictionaries' methods, click the "output" button below to read the full help page, or check out the [official online documentation](https://docs.python.org/3/library/stdtypes.html#dict).

In [172]:
help(dict)

Help on class dict in module builtins:

class dict(object)
 |  dict() -> new empty dictionary
 |  dict(mapping) -> new dictionary initialized from a mapping object's
 |      (key, value) pairs
 |  dict(iterable) -> new dictionary initialized as if via:
 |      d = {}
 |      for k, v in iterable:
 |          d[k] = v
 |  dict(**kwargs) -> new dictionary initialized with the name=value pairs
 |      in the keyword argument list.  For example:  dict(one=1, two=2)
 |  
 |  Methods defined here:
 |  
 |  __contains__(self, key, /)
 |      True if the dictionary has the specified key, else False.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __init__(self,

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 36. Importing and Working with External Libraries

#### Theory
The Python langugage is pretty good, but for a data scientist, the fun begins when you start using the the vast number of high-quality custom libraries that have been written for it.

Some of these libraries are in the "standard library", meaning you can find them anywhere you run Python. Others libraries can be easily added, even if they aren't always shipped with Python.

Either way, we'll access this code with imports.

We'll start our example by importing math from the standard library.

In [131]:
import math

print(f'It\'s math! It has type {type(math)}')

It's math! It has type <class 'module'>


`math` is a module. A module is just a collection of variables (a namespace, if you like) defined by someone else. We can see all the names in `math` using the built-in function `dir()`.

In [132]:
print(dir(math))

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


We can access these variables using dot syntax. Some of them refer to simple values, like `math.pi`:

In [133]:
print('pi to 4 significant digits = {math.pi:.4}')

pi to 4 significant digits = {math.pi:.4}


But most of what we'll find in the module are functions, like `math.log`:

In [134]:
math.log(32, 2)

5.0

Of course, if we don't know what `math.log` does, we can call `help()` on it:

In [135]:
help(math.log)

Help on built-in function log in module math:

log(...)
    log(x, [base=math.e])
    Return the logarithm of x to the given base.
    
    If the base not specified, returns the natural logarithm (base e) of x.



We can also call `help()` on the module itself. This will give us the combined documentation for all the functions and values in the module (as well as a high-level description of the module). Click the "output" button to see the whole `math` help page.

In [136]:
help(math)

Help on module math:

NAME
    math

MODULE REFERENCE
    https://docs.python.org/3.7/library/math
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module provides access to the mathematical functions
    defined by the C standard.

FUNCTIONS
    acos(x, /)
        Return the arc cosine (measured in radians) of x.
    
    acosh(x, /)
        Return the inverse hyperbolic cosine of x.
    
    asin(x, /)
        Return the arc sine (measured in radians) of x.
    
    asinh(x, /)
        Return the inverse hyperbolic sine of x.
    
    atan(x, /)
        Return the arc tangent (measured in radians) of x.
    
    atan2(y, x, /)
        Return the arc tangent (measured in radians) of y/x.
    

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 37. Other import syntax

#### Theory
If we know we'll be using functions in `math` frequently we can import it under a shorter alias to save some typing (though in this case "math" is already pretty short).

In [137]:
import math as mt
mt.pi

3.141592653589793

You may have seen code that does this with certain popular libraries like [pandas](https://pandas.pydata.org/), [NumPy](https://numpy.org/), [Tensorflow](https://www.tensorflow.org/), or [matplotlib](https://matplotlib.org/). For example, it's a common convention to import `numpy as np` and `import pandas as pd`.

The as simply renames the imported module. It's equivalent to doing something like:

In [138]:
import math
mt = math

#### Exercise
...

In [ ]:
# WRITE YOUR CODE HERE

---

## 40. Conclusion
This notebook the basics of working with Python

---

## 41. Bibliography

Learning Python
*    https://www.youtube.com/watch?v=rfscVS0vtbw
*    [Kaggle Learn’s module for Python](https://www.kaggle.com/learn/python). The estimated time of completion is eighthours
*    [Python documentation](https://docs.python.org/3/)
*    [Python Basics](https://fcpython.com/python-basics-fcpython) with [FC Python](https://twitter.com/FC_Python)
*    https://www.w3schools.com/python/default.asp

PES players
*    https://www.instagram.com/unlicensed.fc/

---

***Visit my website [eddwebster.com](https://www.eddwebster.com) or my [GitHub Repository](https://github.com/eddwebster) for more projects. If you'd like to get in contact, my Twitter handle is [@eddwebster](http://www.twitter.com/eddwebster) and my email is: edd.j.webster@gmail.com.***

[Back to the top](#top)